src/net/sourceforge/kolmafia/request/GenericRequest.java

```diff
@@ -1851,18 +1851,21 @@ private boolean handleServerRedirect() {
         return false;
       }
 
-      String oldpwd = GenericRequest.passwordHashValue;
+      String oldpwd = this.getFormField("pwd");
       if (LoginRequest.executeTimeInRequest(this.getURLString(), this.redirectLocation)) {
         if (this.data.isEmpty()) {
-          String newpwd = GenericRequest.passwordHashValue;
-          this.formURLString =
-              StringUtilities.singleStringReplace(this.formURLString, oldpwd, newpwd);
+          if (!oldpwd.equals("")) {
+            String newpwd = GenericRequest.passwordHash;
+            this.formURLString =
+                StringUtilities.singleStringReplace(this.formURLString, oldpwd, newpwd);
+          }
           this.formURL = null;
         } else {
           this.dataChanged = true;
         }
         return false;
       }
+      this.redirectHandled = true;
```

Comment:
Was this deliberately added?

Reply:
Yes.

This happens if executeTimeinRequest returns false. It only does that if it fails to leave you logged in. We return "true", in that case,

Up the call chain, we loop (re)submitting the request until we are told to stop.
Failing to get a response text, getting - and handling - a redirect, and so on.
In particular, handleRedirect returns true if we should NOT resubmit the redirected request, which might have been updated with the redirectLocation.

If executeTimeinRequest returns true, we have successfully logged in, got the new password hash, and - for some reason - visited main.php.In that case, handleServerRedirect returns false - DO resubmit the request which triggered the timein.

If executeTimeinRequest returns false, we did not succeed in logging in. Used to be that only happened if you tried to log in during rollover. But now, since we (might) do a ping test at login, we can fail to log in if the user cancels the ping test, which leaves you logged out. In this case, we return true, which tells the caller to stop submitting the request.

Setting redirectHandled tells run() that even though it got a 302 - from the request that triggered the timein, we we processed it as much as we could and didn't just drop it on the floor. It prevents GenericRequest.run() from reporting "Unhandled redirect". We did handle it. Just not the way we might have hoped.

GenericRequest.run() is really complicated. It is chock full of special cases. Considering that it has grown and mutated for almost 20 years to get into its current state, I'd be scared to try to refactor it.

For good or for ill, my current project is deeply integrated into that code.
Fortunately, I was the one who invented and implemented timing in after KoL times you out.